api_volontaria/apps/volunteer/tests/test_view_events.py

```diff
@@ -260,3 +263,141 @@ def test_list_events(self):
+
+        invalid_json = '{"detail": "extra bracket at the end "}}'
+        response = self.client.post(
+            reverse('event-bulk'),
+            data={
+                "file": BytesIO(),
+                "mapping": invalid_json
+            },
+            format='multipart'
+        )
+
+        content = json.loads(response.content)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(
+            "Mapping should be a dictionary represented in json, errors",
+            content["detail"]
+        )
+
+    def test_bulk_events_mapping_is_not_a_dict(self):
+        """
+        Ensure bad request with explanation is returned if mapping
+        is not a json string representing a dictionary
+        """
+        self.client.force_authenticate(user=self.admin)
+
+        response = self.client.post(
+            reverse('event-bulk'),
+            data={
+                "file": BytesIO(),
+                "mapping": '["list instead of dict"]'
+            },
+            format='multipart'
+        )
+
+        content = json.loads(response.content)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(
+            content,
+            {'detail': "Mapping should be a dictionary pairing "
+                       "the csv column (key) to the element key (value)"}
+        )
+
+    @patch("api_volontaria.apps.volunteer.views.add_bulk_from_file")
+    def test_bulk_events_adding_fail(self, add_bulk_from_file):
+        """
+        Ensure bad request with the caught errors is returned
+        if the bulk operation fails
+        """
+        self.client.force_authenticate(user=self.admin)
+
+        error_message = "error message from add_bulk"
+        add_bulk_from_file.side_effect = InvalidBulkUpdate(
+            error_message
+        )
+
+        response = self.client.post(
+            reverse('event-bulk'),
+            data={"file": BytesIO()},
+            format='multipart'
+        )
+
+        content = json.loads(response.content)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(content, {'detail': error_message})
+
+    @patch("api_volontaria.apps.volunteer.views.add_bulk_from_file")
+    def test_bulk_events_successful(self, add_bulk_from_file):
+        """
+        Ensure 201 created with the created ids is returned
+        if the bulk operation succeeds
+        """
+        self.client.force_authenticate(user=self.admin)
+
+        ids = [4, 56, 89]
+        add_bulk_from_file.return_value = ids
+
+        response = self.client.post(
+            reverse('event-bulk'),
+            data={"file": BytesIO()},
+            format='multipart'
+        )
+
+        content = json.loads(response.content)
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(content, {'created': ids})
```

Comment:
In API conventions we will wait a list of JSON serialized events, a list of events URL or nothing. 

**Option A:** Event url can be retrieve with something like that:

```
reverse(
    'event-detail',
    args=[id_of_event],
)
```

**Option B:** Event serializer format can be output from the serializer you already use.

---

On my opinion, having the serialized format is more useful, but it can fail if somebody try to bulk create 100.000 events.... So maybe just url or nothing should be good.

Reply:
It will return the URLs